New ssa ir

rbartlensky opened the pull request:
I would like some feedback on the current way I am representing instructions [`Instr`](https://github.com/rbartlensky/Lua-interpreter/blob/20b9d623e0e22321493dfddfd3f0e6bc66fb20e5/luacompiler/src/lib/irgen/instr.rs#L49), and SSA opcodes [`IROpcode`](https://github.com/rbartlensky/Lua-interpreter/blob/d6f4f82cacbe28dff39d60c223efca15942f5d43/luacompiler/src/lib/irgen/opcodes.rs).

The whole design of the compiler depends on these two structures, and before moving on to new features I would like to know if they are represented well.

rbartlensky committed:
Refactor the SSA IR.
HLInstr has been removed completely, and replaced by Instr. CompiledFuncs now
have multiple basic blocks, and each basic block has a number of instructions.
This makes it easier to implement branching, and also to construct more complex
instructions.

rbartlensky committed:
Create a compiler from SSA IR to Lua bytecode.
Note that this does not implement register allocation yet!

rbartlensky committed:
Change several instruction implementations.
There are two types of LuaTables:
* `UserTable` which is just a wrapper around HashMap
* `CachingTable` which does faster lookups for strings that are part of the
string constants map.

`UserFunction`s also keep track of a list of upvalues. These are values that
are in scope of a function, but are not defined in the function's scope. The VM
no longer stores `_ENV` in register 0. Instead, it creates a `CachingTable`
upvalue, which is passed to every function as an upvalue. The compiler makes
sure to always store and load global values from the 1st upvalue of a function.

`VarArg` now works very similarly to `MOVR`.

rbartlensky committed:
Add IROpcodes.

rbartlensky committed:
Add if-statements.

rbartlensky committed:
Implement dominators and remove register_map.

rbartlensky committed:
Generate correct Phi instructions.
If outer variables are modified in inner scopes, then the compiler emits Phi
instructions to merge the two values into a new one in order to maintain the
meaning of the original program.

rbartlensky committed:
Generate non-ssa IR by substituting Phis.

rbartlensky committed:
Emit jump instructions.

rbartlensky committed:
Add implementation for Jmp and JmpIf in luavm.

rbartlensky committed:
Refactor the Instr struct.

rbartlensky committed:
Implement while-loops.

rbartlensky committed:
Add less-than operator, and an additional test.

rbartlensky committed:
Implement for-loops.

rbartlensky committed:
Support all relational operators.

rbartlensky committed:
Implement OR and AND short-circuit.
This also fixes a bug in the implementation of CALL. If we have no return
values it should still be possible to MOVR a return value into a register, but
in this case the register will be simply set to Nil.

ltratt commented:
There's no way I can sensibly provide comments on such a big diff without some pointers to guide me: what should I be looking at most of all?

rbartlensky commented:
I should have made it clearer, I only want feedback on the two structs which I linked to in the PR description.

ltratt commented:
OK, got it.

ltratt commented on luacompiler/src/lib/irgen/instr.rs:
```diff
@@ -0,0 +1,104 @@
+}
+
+#[derive(PartialEq, Debug)]
+pub enum Instr {
```
I'm not really sure what this represents: it seems to be the meta-representation of an instruction? Why do we want that vs. having actual instructions (add, sub, mul, call etc.)?

rbartlensky replied:
An instructions has an opcode, and can take zero, one, two, three or n arguments. Before this, I had a struct which had an opcode and a vector of arguments.

ltratt replied:
Yes, but I'm not sure why we're representing instructions at the meta-level. In other words, if I have an instruction `Add(a, b)`, at the meta-level, it's represented as something like `Instr("Add", [Var("a"), Var("b")])`.

rbartlensky replied:
Then would something like:
```rust
enum Instr {
    Add {
        arg1: Arg,
        arg2: Arg,
        arg3: Arg,
    }
    Sub {
        arg1: Arg,
        arg2: Arg,
        arg3: Arg,
    }
    ...
}
```
be more suitable?

ltratt replied:
I'm not saying which is more suitable, but I am asking you which is more suitable :)

rbartlensky replied:
I am not sure which is more suitable. I would say my current solution involves less code duplication. If I were to use the above representation then I would have a need similar structs for all arithmetic operators. I could create a structure to represent an arithmetic operation, but I think  that will lead to unnecessary indirection.

ltratt replied:
The honest answer is that I don't know and although my gut feeling vaguely suggests the opposite, it's not a very strong gut feeling. So I suggest you carry on down this route and see where it leads you!

jacob-hughes replied:
I just saw this in the PR channel and thought I'd chime in and say it might be worth looking at how rustc represents it's MIR (Mid-Level IR). MIR isn't in SSA form, but I don't think that's particularly relevant to this comment.

https://github.com/rust-lang/rust/blob/1349c84a4fa0fca9b866b2e859d28ee185ca0c1b/src/librustc/mir/mod.rs#L2190

They don't create instructions at the meta-level but they do create a single BinaryOp instruction and parameterize the actual binary operator. It might be overkill for your situation though

rbartlensky replied:
I see, thank you! I will stick with my current design for now and see where it takes me. If I decide to switch to explicit instructions, I will probably represent binary operations in a similar way.

ltratt commented on luacompiler/src/lib/irgen/opcodes.rs:
```diff
@@ -0,0 +1,23 @@
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum IROpcode {
+    Opcode(Opcode),
+    Phi,
```
Naively, I was expecting this to have one or more parents node pointers?

rbartlensky replied:
The basic blocks record their parents, and phis are generated based on that.

ltratt replied:
OK.

rbartlensky committed:
Refactor IROpcodes.

rbartlensky committed:
Implement negative ranges for for-loops.
Also fix AND short-circuit logic.

rbartlensky committed:
Add support for `prefixexp` in luacompiler.

rbartlensky committed:
Load the io module.

rbartlensky committed:
Add support for command-line arguments.

rbartlensky committed:
Add lua_tonumber.

rbartlensky committed:
Handle upvalues in luacompiler.

ltratt commented:
I think we can merge this?

rbartlensky committed:
Fix call stack bug, and improve performance.
In order to illustrate the bug, consider the following program:
```
1. function f()
2.   print(2)
3.   return 1
4. end
5. print(f())
```
The vm prepares the stack for the call to `print` at line 5. Then it prepares
the stack for the call to `f`. Each function has a `closure` object associated
to it, and this was used to keep track of where the arguments are on the stack.
So the vm sets `print`'s `args_start` to whatever the top of the stack was
before the call to `f`, but `f` sets `args_start` for `print` as well, without
restoring the old value. This was solved by tracking each stack frame in the vm,
instead of associating the beginning of a frame with each closure.

rbartlensky commented:
I will change this to a regular PR in order to merge it, and then I will continue with feature branches. if that is fine with you.

ltratt commented:
Sounds like a plan!
